### PR TITLE
WIP: Multi-testnets 

### PR DESCRIPTION
**Pull request by utx0**
## Description

WIP: Only partially working. 

---

### Author Checklist

_All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues._

I have...

-   [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
-   [ ] added `!` to the type prefix if API or client breaking change
-   [ ] provided a link to the relevant issue or specification
-   [ ] included the necessary unit and integration tests
-   [ ] code is adequately commented
-   [ ] updated the relevant documentation or specification
-   [ ] reviewed "Files changed" and left comments if necessary
-   [ ] confirmed all CI checks have passed

### Reviewers Checklist

_All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items._

I have...

-   [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
-   [ ] confirmed `!` in the type prefix if API or client breaking change
-   [ ] confirmed all author checklist items have been addressed
-   [ ] reviewed service design and naming
-   [ ] reviewed documentation is accurate
-   [ ] reviewed tests and test coverage
-   [ ] manually tested (if applicable)

### CODE REVIEW

File: proposals/Addresses.sol [hunk 1]
```diff
@@ -4,126 +4,251 @@ pragma solidity 0.8.18;
...
+
+enum EnvVar {
+    LocalNet,
+    SandPitNet, // Tuxedo's team env for localized testing on sepolia
+    DevNet, // dev env
+    TestNet, // qa env
+    MixedNet, // TODO to be removed. Only used for contracts getting used in devNet and testNet
+    StageNet, // staging env
+    MainNet // prod env
+}
```
Comment (ElliotFriedman) [inline proposals/Addresses.sol:17]: why do you have to use an enum? You'll have to convert this to a number when you change this structure to read in addresses from a json file.

seems way simpler to just make a set of strings, and then use the strings in addresses.sol, then if a string doesn't match the pre-specified strings you just throw an error

File: proposals/Addresses.sol [hunk 2]
```diff
@@ -4,126 +4,251 @@ pragma solidity 0.8.18;
...
+        return EnvVar.SandPitNet;
+    } else if (31337 == chainId) {
+        return EnvVar.LocalNet;
+    } else if (421614 == chainId) {
+        return EnvVar.DevNet;
+    } else if (421614 == chainId) {
+        return EnvVar.TestNet;
+    } else if (421614 == chainId) {
+        return EnvVar.MixedNet;
+    } else if (421614 == chainId) {
```
Comment (ElliotFriedman) [inline proposals/Addresses.sol]: this doesn't make any sense, you can't tell the env var from the chainid

File: proposals/TestProposals.sol [hunk 3]
```diff
@@ -41,7 +41,7 @@ contract TestProposals is Test {
     address public deployer = address(0x00000108);
 
     function setUp() public {
-        addresses = new Addresses();
+        addresses = new Addresses(EnvVar.MainNet); // TODO how do I make this configurable from the fork in use?
```
Comment (ElliotFriedman) [inline proposals/TestProposals.sol:44]: you have to change the addresses.sol contract to use the chainid to determine which fork you're on and read in all of the addresses that exist on that chain.

then I'd add another function outside of the constructor where you can set the env of the network you're on.

File: proposals/Addresses.sol [hunk 4]
```diff
@@ -4,126 +4,251 @@ pragma solidity 0.8.18;
...
+        } else if (EnvVar.TestNet == envVar) {
+            setTestNetAddresses();
+        } else if (EnvVar.DevNet == envVar) {
+            setDevNetAddresses();
+        } else if (EnvVar.LocalNet == envVar) {
+            setLocalNetAddresses();
+        } else if (EnvVar.MixedNet == envVar) {
+            setMixedNetAddresses();
+        } else if (EnvVar.SandPitNet == envVar) {
+            setSandPitNetAddresses();
```
Comment (ElliotFriedman) [inline proposals/Addresses.sol:101]: you should read all of these into storage by default

File: proposals/Addresses.sol [hunk 5]
```diff
@@ -4,126 +4,251 @@ pragma solidity 0.8.18;
...
+    //     }
+
+    //     console.log("Fork RPC: ", fork);
+    //     // uint x = vm.createFork(fork);
+    //     // vm.selectFork(x);
+
+    //     vm.createSelectFork(fork);
+    // }
+
     function _addAddress(string memory name, address addr) private {
```
Comment (ElliotFriedman) [inline proposals/Addresses.sol:233]: I'd add another version of this function that takes the env var as a param so you can load up all of the chains when you spin up addresses.sol

File: proposals/Addresses.sol [hunk 6]
```diff
@@ -4,126 +4,251 @@ pragma solidity 0.8.18;
...
+    constructor(EnvVar _env) {
         chainId = block.chainid;
+        envVar = _env;
+
         console.log("chainId: ", chainId);
+        console.log("envVar: ", getEnvVar(envVar));
+
+        setNetworkAddresses(envVar);
+    }
+
```
Comment (ElliotFriedman) [inline proposals/Addresses.sol:84]: I would add a helper function here that allows you to set the state variable `envVar ` which is used for retrieving items from the mapping. That way it's easy to flip through networks and devnets if you need to without restarting processes.

File: proposals/Addresses.sol [hunk 7]
```diff
@@ -4,126 +4,251 @@ pragma solidity 0.8.18;
 import {Test} from "@forge-std/Test.sol";
 import {console} from "@forge-std/console.sol";
```
Comment (ElliotFriedman) [inline proposals/Addresses.sol:6]: the new version of addresses.sol includes a bunch of null address checks which can really save you if you ask for an address that isn't there in a deployment script, and it allows CRUD. https://github.com/solidity-labs-io/forge-proposal-simulator/blob/main/addresses/Addresses.sol

File: proposals/Addresses.sol (hunk 1 above)
Comment (utx0) [reply proposals/Addresses.sol:17]: string as not type safe. Go learn about how to program in type safe language and their well known patterns.

File: proposals/Addresses.sol (hunk 2 above)
Comment (utx0) [reply proposals/Addresses.sol]: As this code isn't ready for review it still have things I need to fix and change and this most likely be removed.

File: proposals/Addresses.sol (hunk 6 above)
Comment (utx0) [reply proposals/Addresses.sol:84]: Sorry I dont follow, pls give code example.

File: proposals/Addresses.sol (hunk 4 above)
Comment (utx0) [reply proposals/Addresses.sol:101]: Not sure I follow? This is just picking which network set of addresses to load. Nothing about the loading of these values has changed.

File: proposals/TestProposals.sol (hunk 3 above)
Comment (utx0) [reply proposals/TestProposals.sol:44]: This was a question for me to come back to. Pls dont waste your time and mine with a PR that is not ready for review as you often miss the point of where things are going.

File: proposals/Addresses.sol (hunk 5 above)
Comment (utx0) [reply proposals/Addresses.sol:233]: why would we want/need that?

**Pull request closed without merging**